caffe2/serialize/inline_container.cc

```diff
@@ -0,0 +1,305 @@
+  size_t mod = start % kFieldAlignment;
+  size_t next_offset = (mod == 0) ? start : (start + kFieldAlignment - mod);
+  size_t padding_size = next_offset - start;
+  std::string buf(padding_size + 4, 'Z');
+  // zip extra encoding (key, size_of_extra_bytes)
+  buf[0] = 'F';
+  buf[1] = 'B';
+  buf[2] = (uint8_t) padding_size;
+  buf[3] = (uint8_t) (padding_size >> 8);
+  return buf;
+}
+
+size_t PyTorchStreamReader::getFileID(const std::string& name) {
+  std::stringstream ss;
+  ss << archive_name_ << "/" << name;
+  size_t result = mz_zip_reader_locate_file(ar_.get(), ss.str().c_str(), nullptr, 0);
+  if (ar_->m_last_error == MZ_ZIP_FILE_NOT_FOUND) {
+    CAFFE_THROW("file not found: ", ss.str());
+  }
+  valid("locating file");
+  return result;
+}
+
+// return dataptr, size
+std::tuple<at::DataPtr, size_t> PyTorchStreamReader::getRecord(const std::string& name) {
+  size_t key = getFileID(name);
+  mz_zip_archive_file_stat stat;
+  mz_zip_reader_file_stat(ar_.get(), key, &stat);
+  valid("retrieving file meta-data");
+  void * ptr = malloc(stat.m_uncomp_size);
+  mz_zip_reader_extract_to_mem(ar_.get(), key, ptr, stat.m_uncomp_size, 0);
+  valid("reading file");
+
+  at::DataPtr retval(ptr, ptr, free, at::kCPU);
+  return std::make_tuple(std::move(retval), stat.m_uncomp_size);
+}
+
+static int64_t read_le_16(uint8_t* buf) {
+  return buf[0] + (buf[1] << 8);
+}
+
+size_t PyTorchStreamReader::getRecordOffset(const std::string& name) {
+  mz_zip_archive_file_stat stat;
+  mz_zip_reader_file_stat(ar_.get(), getFileID(name), &stat);
+  valid("retriving file meta-data");
+  in_->seekg(stat.m_local_header_ofs);
+  valid("seeking to file header");
+  uint8_t local_header[MZ_ZIP_LOCAL_DIR_HEADER_SIZE];
+  in_->read(reinterpret_cast<char*>(local_header), MZ_ZIP_LOCAL_DIR_HEADER_SIZE);
+  valid("reading file header");
+  size_t filename_len = read_le_16(local_header + MZ_ZIP_LDH_FILENAME_LEN_OFS);
+  size_t extra_len = read_le_16(local_header + MZ_ZIP_LDH_EXTRA_LEN_OFS);
+  return stat.m_local_header_ofs + MZ_ZIP_LOCAL_DIR_HEADER_SIZE + filename_len + extra_len;
+}
+
+
+PyTorchStreamReader::~PyTorchStreamReader() {
+  mz_zip_reader_end(ar_.get());
+  valid("closing reader");
+}
+
+size_t ostream_write_func(void *pOpaque, mz_uint64 file_ofs, const void *pBuf, size_t n) {
+  auto self = static_cast<PyTorchStreamWriter*>(pOpaque);
+  if (self->current_pos_ != file_ofs) {
+    // xxx - windows ostringstream refuses to seek to the end of an empty string
+    // so we workaround this by not calling seek unless necessary
+    // in the case of the first write (to the empty string) file_ofs and
+    // current_pos_ will be 0 and the seek won't occur.
+    self->out_->seekp(file_ofs);
+    if(!*self->out_)
+      return 0;
+  }
+
+  self->out_->write(static_cast<const char*>(pBuf), n);
+  if(!*self->out_)
+    return 0;
+  self->current_pos_ = file_ofs + n;
+  return n;
+}
+
+PyTorchStreamWriter::PyTorchStreamWriter(std::string file_name, std::ostream* out)
```

Comment:
Is there a reason we're not passing `file_name` by const ref?

Reply:
no, at some point we were universally capturing it, and I didn't update the API.